In [1]:
import numpy as np
import pandas as pd

1. Load the dataset and handle any missing data by replacing null values with the mean
value of the respective column. (Score 2)

In [6]:
df = pd.read_csv('wineq.csv',sep=';')
df.head()

,fixed acidity,volatile acidity,citric acid,residual sugar,chlorides,free sulfur dioxide,total sulfur dioxide,density,pH,sulphates,alcohol,quality
0,7.4,0.70,0.00,1.9,0.076,11.0,34.0,0.9978,3.51,0.56,9.4,5
1,7.8,0.88,0.00,2.6,0.098,25.0,67.0,0.9968,3.20,0.68,9.8,5
2,7.8,0.76,0.04,2.3,0.092,15.0,54.0,0.9970,3.26,0.65,9.8,5
3,11.2,0.28,0.56,1.9,0.075,17.0,60.0,0.9980,3.16,0.58,9.8,6
4,7.4,0.70,0.00,1.9,0.076,11.0,34.0,0.9978,3.51,0.56,9.4,5


In [10]:
print(df.isnull().sum())
print("  ")
df= df.fillna(df.mean(numeric_only=True))
print(df.isnull().sum())

fixed acidity           0
volatile acidity        0
citric acid             0
residual sugar          0
chlorides               0
free sulfur dioxide     0
total sulfur dioxide    0
density                 0
pH                      0
sulphates               0
alcohol                 0
quality                 0
dtype: int64
  
fixed acidity           0
volatile acidity        0
citric acid             0
residual sugar          0
chlorides               0
free sulfur dioxide     0
total sulfur dioxide    0
density                 0
pH                      0
sulphates               0
alcohol                 0
quality                 0
dtype: int64


There are no missing values in this dataset, so no values actually needed to be replaced.

2. Extract the following columns as vectors: alcohol,citric acid. (Score 2)

In [13]:
alcohol = df["alcohol"].to_numpy()
citric = df["citric acid"].to_numpy()

In [ ]:
print("Alcohol vector : ")
print(alcohol)

print("\nCitricACid Vector: ")
print(citric)

Alcohol vector : 
[ 9.4  9.8  9.8 ... 11.  10.2 11. ]

CitricACid Vector: 
[0.   0.   0.04 ... 0.13 0.12 0.47]


3. Select two features (e.g., alcohol and density) from the dataset and calculate the
covariance matrix using np.cov(X.T), where X is the feature matrix consisting of the
selected columns.(Score 2)

In [16]:
X = df[["alcohol","density"]]
covmatrix = np.cov(X.T)

print("Covariance Matrix: ")
print(covmatrix)

Covariance Matrix: 
[[ 1.13564740e+00 -9.97951790e-04]
 [-9.97951790e-04  3.56202945e-06]]


The diagonal values represent the variance of alcohol and density.
The off-diagonal values represent the covariance between alcohol and density.
The negative covariance indicates a slight negative relationship between alcohol and density.


4. Perform eigen decomposition on the covariance matrix you computed in question 3.
Identify and interpret the results:Identify the top 2 eigenvalues of the covariance
matrix,Identify the corresponding eigenvectors.(Score 2)

In [17]:
eigenvalues, eigenvectors = np.linalg.eig(covmatrix)
print("Eigenvalues:")
print(eigenvalues)

print("\nEigenvectors:")
print(eigenvectors)

Eigenvalues:
[1.13564827e+00 2.68507580e-06]

Eigenvectors:
[[ 9.99999614e-01  8.78753184e-04]
 [-8.78753184e-04  9.99999614e-01]]


In [18]:
index = np.argsort(eigenvalues)[::-1]

eigenvalues = eigenvalues[index]
eigenvectors = eigenvectors[:, index]

print("Top 2 Eigenvalues:")
print(eigenvalues)

print("\nCorresponding Eigenvectors:")
print(eigenvectors)

Top 2 Eigenvalues:
[1.13564827e+00 2.68507580e-06]

Corresponding Eigenvectors:
[[ 9.99999614e-01  8.78753184e-04]
 [-8.78753184e-04  9.99999614e-01]]


The first eigenvalue is much larger than the second one. This means that most of the variation in the selected features (alcohol and density) is along the direction represented by the first eigenvector.

The first eigenvector is mainly associated with alcohol, because its value for alcohol is almost 1.

5. Which wine quality is most common in the dataset? How can you interpret the
distribution of wine quality scores? (Score 2)

In [20]:
quality_counts = df["quality"].value_counts().sort_index()
print(quality_counts)

quality
3     10
4     53
5    681
6    638
7    199
8     18
Name: count, dtype: int64


In [22]:
print("Most common wine quality: ",quality_counts.idxmax())
print("Number of Wines: ",quality_counts.max())

Most common wine quality:  5
Number of Wines:  681


The most common wine quality is 5, with 681 wines.



Interpretation:

The distribution is concentrated mainly around quality scores 5 and 6. Very few wines have quality scores of 3, 4, or 8. Therefore, most wines in this dataset have average or moderate quality, while very low and very high quality wines are less common.